docroot/modules/custom/mass_views/src/Plugin/Action/AddCollectionsDocuments.php

```diff
@@ -0,0 +1,210 @@
+      return $return_as_object ? $access : $access->isAllowed();
+    }
+
+    // Other entity types may have different
+    // access methods and properties.
+    return TRUE;
+  }
+
+  /**
+   * Returns the entity bundles allowed for collections.
+   */
+  private function intersectTargetBundles() {
+    $media_storage = \Drupal::entityTypeManager()->getStorage('media');
+    $target_bundles = NULL;
+
+    /** @var int[] */
+    $list = $this->context['list'];
+    foreach ($list as $item_id) {
+      $media = $media_storage->load(array_reverse($item_id)[0]);
+      if (!empty($media)) {
+        /** @var \Drupal\entity_hierarchy\Plugin\Field\FieldType\EntityReferenceHierarchyFieldItemList */
+        $collections = $media->hasField('field_collections') ?? FALSE;
+        $definition = $collections ? $media->field_collections->getFieldDefinition() : FALSE;
+        $settings = $definition ? $definition->getSettings() : FALSE;
+        $handler_settings = $settings ? $settings['handler_settings'] ?? [] : [];
+        $target_bundles =
+          is_array($target_bundles) ?
+            \array_intersect($target_bundles, ($handler_settings['target_bundles'] ?? [])) :
+            $handler_settings['target_bundles'];
+      }
+    }
+
+    return $target_bundles;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    $target_bundles = $this->intersectTargetBundles();
+    if (!empty($target_bundles)) {
+
+      $vocabularies = Vocabulary::loadMultiple($target_bundles);
+
+      $form['#list'] = $this->context['list'];
+
+      $form['actions']['submit']['#value'] = $this->t('Add collections');
+
+      $form['new_collection'] = [
+        '#type' => 'checkbox_tree',
+        '#vocabularies' => $vocabularies,
+        '#max_choices' => -1,
+        '#leaves_only' => FALSE,
+        '#select_parents' => TRUE,
+        '#cascading_selection' => 0,
+        '#value_key' => 'target_id',
+        '#max_depth' => 0,
+        '#start_minimized' => TRUE,
+        '#title' => $this->t('New Collection'),
+        '#required' => TRUE,
+        '#attributes' => ['class' => ['field--widget-term-reference-tree']]
+      ];
+    }
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  private function getValueFromElement($element, FormStateInterface $form_state) {
+    $items = _term_reference_tree_flatten($element, $form_state);
+    $value = [];
+    if ($element['#max_choices'] != 1) {
+      foreach ($items as $child) {
+        if (!empty($child['#value'])) {
+          // If the element is leaves only and select parents is on,
+          // then automatically add all the parents of each selected value.
+          if (!empty($element['#select_parents']) && !empty($element['#leaves_only'])) {
+            foreach ($child['#parent_values'] as $parent_tid) {
+              if (!in_array([$element['#value_key'] => $parent_tid], $value)) {
```

Comment:
I often go by a rule of "limit nested blocks to 3 levels", and if you have more it's a signal that a new private method should be refactored out. I don't see it as a blocker for this PR, unless we end up having to change these lines anyways.

Reply:
I think I've grabbed the code form the contrib module,I didn't think of that. Do you think it's critical and worth attention?

Reply:
No, I would say leave it at this point unless we're making changes to these lines. But, let's keep it in mind for future work?